File: src/main/java/xyz/oribuin/xpbottles/command/BottleCommand.java
package xyz.oribuin.xpbottles.command;

import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import xyz.oribuin.xpbottles.XPBottlePlugin;
import xyz.oribuin.xpbottles.util.HexUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BottleCommand implements TabExecutor {

    private final XPBottlePlugin plugin;

    public BottleCommand(@NotNull XPBottlePlugin plugin) {
        this.plugin = plugin;

        final PluginCommand command = this.plugin.getCommand("bottle");
        if (command != null) {
            command.setExecutor(this);
            command.setTabCompleter(this);
        }

    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {

        // Invalid usage
        if (args.length == 0) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-args")));
            return true;
        }

        switch (args[0].toLowerCase()) {
            case "reload":
                this.reload(sender);
                break;

            case "give":
                this.give(sender, args);
                break;

            case "max":
                if (!(sender instanceof Player)) {
                    sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("player-only")));
                    return true;
                }

                this.bottle(sender, Math.min((this.plugin.getTotalXP((Player) sender)), this.plugin.getConfig().getInt("max-xp")));
            default:

                if (!(sender instanceof Player)) {
                    sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("player-only")));
                    return true;
                }

                int amount;

                try {
                    amount = Integer.parseInt(args[0]);
                } catch (NumberFormatException ex) {
                    sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-amount")));
                    return true;
                }

                this.bottle(sender, amount);
                break;
        }

        return true;
    }

    /**
     * Create a bottle with the given xp and owner
     *
     * @param sender The player
     * @param amount The amount of xp to give
     */
    private void bottle(CommandSender sender, int amount) {
        if (!sender.hasPermission("xpbottles.bottle")) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("no-permission")));
            return;
        }

        final Player player = (Player) sender;
        final ItemStack result = this.plugin.createBottle(sender, amount);

        // Make sure amount is valid
        if (amount < this.plugin.getConfig().getInt("min-xp") || amount > this.plugin.getConfig().getInt("max-xp")) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-amount")));
            return;
        }

        int playerXP = this.plugin.getTotalXP(player);
        if (amount > playerXP) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("not-enough-xp")));
            return;
        }

        // Inventory is full
        if (player.getInventory().firstEmpty() == -1) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("full-inventory")));
            player.getWorld().dropItem(player.getLocation(), result,
                    item -> item.getPersistentDataContainer()
                            .set(
                                    this.plugin.getOwnerKey(),
                                    PersistentDataType.STRING,
                                    player.getUniqueId().toString()
                            ));

            return;
        }

        // give bottle ^.^
        player.getInventory().addItem(result);
    }

    /**
     * Give command for the plugin
     *
     * @param sender The command sender
     * @param args   The command arguments
     */
    private void give(CommandSender sender, String[] args) {

        // no perms :(
        if (!sender.hasPermission("xpbottles.give")) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("no-permission")));
            return;
        }

        // invalid usage
        if (args.length < 3) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-args")));
            return;
        }

        // invalid player
        final Player player = this.plugin.getServer().getPlayer(args[1]);
        if (player == null) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-player")));
            return;
        }

        // invalid amount
        int xpAmount;
        try {
            xpAmount = Integer.parseInt(args[2]);
        } catch (NumberFormatException ex) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-args")));
            return;
        }

        // invalid amount 2 electric boogaloo
        if (xpAmount <= 0) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("invalid-args")));
            return;
        }

        final ItemStack result = this.plugin.createBottle(player, xpAmount);

        // Inventory is full
        if (player.getInventory().firstEmpty() == -1) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("full-inventory")));

            // Drop the bottle on the floor
            player.getWorld().dropItem(player.getLocation(), result,
                    item -> item.getPersistentDataContainer()
                            .set(
                                    this.plugin.getOwnerKey(),
                                    PersistentDataType.STRING,
                                    player.getUniqueId().toString()
                            ));
            return;
        }

        // give bottle ^.^
        player.getInventory().addItem(result);
    }

    /**
     * Reload command for the plugin
     *
     * @param sender The command sender
     */
    private void reload(CommandSender sender) {

        if (!sender.hasPermission("xpbottles.reload")) {
            sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("no-permission")));
            return;
        }

        this.plugin.reload();
        sender.sendMessage(HexUtils.colorify(this.plugin.getConfig().getString("reload")));
    }

    @Nullable
    @Override
    public List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {
        // TODO: Improve tab completion
        return Arrays.asList("reload", "give", "max", "<amount>");
    }

}

File: src/main/java/xyz/oribuin/xpbottles/command/BottleCommand.kt
package xyz.oribuin.xpbottles.command

import org.bukkit.NamespacedKey
import org.bukkit.command.CommandSender
import org.bukkit.entity.Item
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataType
import xyz.oribuin.orilibrary.command.Command
import xyz.oribuin.orilibrary.util.StringPlaceholders
import xyz.oribuin.xpbottles.XPBottles
import xyz.oribuin.xpbottles.manager.MessageManager

@Command.Info(
    name = "bottle",
    aliases = ["xpbottle"],
    description = "Main command for obtaining xp bottles",
    usage = "/bottle <amount>",
    permission = "xpbottle.use",
    playerOnly = false,
    subCommands = [GiveSub::class, ReloadSub::class]
)
class BottleCommand(private val plugin: XPBottles) : Command(plugin) {

    private val msg = this.plugin.getManager(MessageManager::class.java)

    override fun runFunction(sender: CommandSender, label: String, args: Array<String>) {

        if (args.isEmpty()) {
            this.msg.send(sender, "invalid-args", StringPlaceholders.single("usage", "/bottle <amount>"))
            return
        }

        val subCommand = this.subCommands.find { subCommand -> subCommand.info.names.contains(args[0].lowercase()) }

        // Check if the sender used a subcommand
        if (subCommand != null) {
            this.runSubCommands(sender, args, { this.msg.send(it, "unknown-cmd") }, { this.msg.send(it, "no-perm") })
            return
        }

        // Check if sender is a player, console cant withdraw xp
        if (sender !is Player) {
            this.msg.send(sender, "player-only")
            return
        }

        val amount = if (args[0].equals("all", ignoreCase = true) || args[0].equals("max", ignoreCase = true))
            sender.totalExperience.coerceAtMost(plugin.config.getInt("max-xp"))
        else
            args[0].toIntOrNull()

        // Check if the player put in a valid number or if their number is less or equal to 0
        if (amount == null || amount <= 0) {
            this.msg.send(sender, "invalid-amount")
            return
        }

        // Check if the sender has enough xp to do it
        if (plugin.getTotalXp(sender) < this.plugin.config.getInt("min-xp")) {
            this.msg.send(sender, "not-enough-xp")
            return
        }

        // Check if the amount provided is more than cap
        if (amount > this.plugin.config.getInt("max-xp")) {
            this.msg.send(sender, "max-xp")
            return
        }

        if (amount < this.plugin.config.getInt("min-xp")) {
            this.msg.send(sender, "min-xp")
            return
        }

        if(plugin.getTotalXp(sender) < amount) {
            this.msg.send(sender, "not-enough-xp")
            return
        }

        val bottle = this.plugin.createBottle(amount, sender)
        plugin.setXp(sender, plugin.getTotalXp(sender) -amount)
        this.msg.send(sender, "withdrew-xp", StringPlaceholders.single("xp", plugin.formatNum(amount)))
        if (sender.inventory.firstEmpty() == -1) {
            this.msg.send(sender, "dropped-xp-bottle")

            sender.world.spawn(sender.location, Item::class.java) {
                it.itemStack = bottle
                it.persistentDataContainer.set(NamespacedKey(plugin, "bottleOwner"), PersistentDataType.STRING, sender.uniqueId.toString())
            }

            return
        }

        sender.inventory.addItem(bottle)
    }

    override fun completeString(sender: CommandSender, label: String, args: Array<String>): MutableList<String> {
        val tabComplete = mutableListOf<String>()

        when (args.size) {
            1 -> {
                tabComplete.addAll(subCommands.filter { sender.hasPermission(it.info.permission) }.map { it.info.names[0].lowercase() })
                tabComplete.addAll(listOf("<amount>", "max", "all"))
            }

            2 -> if (args[0].equals("give", true))
                tabComplete.addAll(playerList(sender))

            3 -> if (args[0].equals("give", true))
                tabComplete.add("<amount>")
        }
        return tabComplete
    }

}